src/sage/rings/semirings/tropical_polynomial.py

```diff
@@ -0,0 +1,994 @@
+
+            sage: p2 = R([4,4,2]); p2
+            2*x^2 + 4*x + 4
+            sage: p2.factor()
+            (2) * (0*x^2 + 2*x + 2)
+
+        TESTS:
+
+        The factorization for a constant::
+
+            sage: p3 = R(3)
+            sage: p3.factor()
+            (3) * 0
+        """
+        from sage.structure.factorization import Factorization
+        unit = self.dict()[self.degree()]
+        if (self != self.split_form()) or (not self.roots()):
+            factor = [(self * self.parent(-unit.lift()), 1)]
+            return Factorization(factor, unit=unit)
+
+        R = self.parent()
+        roots_order = {}
+        for root in self.roots():
+            if root in roots_order:
+                roots_order[root] += 1
+            else:
+                roots_order[root] = 1
+        factors = []
+        for root in roots_order:
+            factors.append((R([root, 0]), roots_order[root]))
+
+        return Factorization(factors, unit=unit)
+
+    def piecewise_function(self):
+        r"""
+        Return the tropical polynomial function of ``self``.
+
+        The function is a piecewise linear function with the domains are
+        divided by the roots. First we convert each term of polynomial to
+        its corresponding linear function. Next, we must determine which
+        term achieves the minimum (maximum) at each interval.
+
+        OUTPUT: A piecewise function
+
+        EXAMPLES::
+
+            sage: T = TropicalSemiring(QQ, use_min=False)
+            sage: R.<x> = PolynomialRing(T)
+            sage: p1 = R([4,2,1,3]); p1
+            3*x^3 + 1*x^2 + 2*x + 4
+            sage: p1.piecewise_function()
+            piecewise(x|-->4 on (-oo, 1/3], x|-->3*x + 3 on (1/3, +oo); x)
+
+        A constant tropical polynomial will result in a constant function::
+
+            sage: p2 = R(3)
+            sage: p2.piecewise_function()
+            3
+
+        A monomial will resulted in a linear function::
+
+            sage: p3 = R(1)*x^3
+            sage: p3.piecewise_function()
+            3*x + 1
+        """
+        from sage.symbolic.ring import SR
+        from sage.functions.piecewise import piecewise
+        from sage.sets.real_set import RealSet
+
+        x = SR.var('x')
+        data = self.dict()
+        R = self.base_ring()
+        if not self.roots():
+            f = data[0].lift()
+            return f
+
+        if len(data) == 1:
+            gradient = list(data)[0]
+            intercept = data[gradient].lift()
+            f = intercept + gradient * x
+            return f
+
+        unique_root = sorted(list(set(self.roots())))
+        pieces = []
+        domain = []
+        for i in range(len(unique_root)+1):
+            if i == 0:
+                test_number = R(unique_root[i]-1)
+            elif i == len(unique_root):
+                test_number = R(unique_root[i-1]+1)
+            else:
+                test_number = R((unique_root[i]+unique_root[i-1])/2)
```

Comment:
```suggestion
                test_number = R(unique_root[i] - 1)
            elif i == len(unique_root):
                test_number = R(unique_root[i-1] + 1)
            else:
                test_number = R((unique_root[i]+unique_root[i-1]) / 2)
```
PEP8 changes need to be done.